File: clang/test/OmpSs-RT/07_phases_ompss.dg/c/success_task_08.c
/*--------------------------------------------------------------------
  (C) Copyright 2006-2012 Barcelona Supercomputing Center
                          Centro Nacional de Supercomputacion
  
  This file is part of Mercurium C/C++ source-to-source compiler.
  
  See AUTHORS file in the top level directory for information
  regarding developers and contributors.
  
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.
  
  Mercurium C/C++ source-to-source compiler is distributed in the hope
  that it will be useful, but WITHOUT ANY WARRANTY; without even the
  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
  PURPOSE.  See the GNU Lesser General Public License for more
  details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with Mercurium C/C++ source-to-source compiler; if
  not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.
--------------------------------------------------------------------*/

// RUN: %oss-compile-and-run
// RUN: %oss-O2-compile-and-run

/*
<testinfo>
test_generator=(config/mercurium-ompss "config/mercurium-ompss-2 openmp-compatibility")
</testinfo>
*/

#include<assert.h>

void foo(int* n)
{
    *n = 2;
}

#pragma oss task inout(*n) final(1)
void bar(int* n)
{
    *n = 4;
}

int main()
{
    int x = -1;
#pragma oss task inout(x) final(1)
    {
        foo(&x);
    }

#pragma oss taskwait on(x)
    assert(x == 2);

    bar(&x);
#pragma oss taskwait on(x)
    assert(x == 4);

    return 0;
}

File: flang/test/OmpSs-RT/07_phases_ompss.dg/fortran/success_task_08.f90
! RUN: %oss-compile-and-run
! XFAIL: true
! <testinfo>
! test_generator=(config/mercurium-ompss "config/mercurium-ompss-2 openmp-compatibility")
! </testinfo>
PROGRAM MAIN
  IMPLICIT NONE

  INTEGER :: A1, B1, A2, B2, I, J
  INTEGER, DIMENSION(:, :), ALLOCATABLE :: X, Y

  ALLOCATE(X(2:100, 3:75), Y(2:100, 3:75))

  A1 = 5
  B1 = 10
  A2 = 6
  B2 = 11

  Y = 42

  !$OSS TASK OUT(X(A1:B1, A2:B2))
  X(5:10, 6:11) = 3
  !$OSS END TASK

  A1 = 5+1
  B1 = 10+1
  A2 = 6+1
  B2 = 11+1

  !$OSS TASK OUT(Y(A1-1:B1-1, A2-1:B2-1)) IN(X(A1-1:B1-1, A2-1:B2-1))
  Y(5:10, 6:11) = X(5:10,6:11) + 1
  !$OSS END TASK

  !$OSS TASKWAIT

  DO I = 5, 10
   DO J = 6, 11
     IF (5 <= I &
         .AND. I <= 10 &
         .AND. 6 <= J &
         .AND. J <= 11) THEN
        IF (Y(I, J) /= 4) STOP 1
     ELSE
        IF (Y(I, J) /= 42) STOP 2
     END IF
   END DO
  END DO
END PROGRAM MAIN

